tools/cmar.ml

```diff
@@ -0,0 +1,225 @@
+          )
+          else
+            (missing_in_library, library_rest)
+      )
+      ([], library_units) units_to_extract)
+;;
+
+type unit_type = CMOUnit of Cmo_format.compilation_unit | 
+                 CMXUnit of Cmx_format.unit_infos * Digest.t;;
+
+(* Extract single unit from library. *)
+let extract_unit (library_ichannel, library_info) unit_name =
+  let comp_unit, unit_file_ext =
+    match library_info with
+    | CMAlibrary lib ->
+      let comp_unit = 
+        List.find 
+          (fun cmo_unit -> cmo_unit.Cmo_format.cu_name = unit_name) 
+          lib.Cmo_format.lib_units
+      in
+      (CMOUnit comp_unit, ".cmo")
+    | CMXAlibrary lib ->
+      let comp_unit, digest =
+        List.find 
+          (fun (cmx_unit, _) -> cmx_unit.Cmx_format.ui_name = unit_name) 
+          lib.Cmx_format.lib_units
+      in
+      (CMXUnit (comp_unit, digest), ".cmx")
+  in
+  let unit_ochannel = 
+    open_out_bin ((String.uncapitalize_ascii unit_name) ^ unit_file_ext)
+  in
+  (
+    match comp_unit with
+    | CMOUnit comp_unit ->
+      output_string unit_ochannel Config.cmo_magic_number;
+      let ofs_pos_info = pos_out unit_ochannel in
+      output_binary_int unit_ochannel 0;
+
+      (* copying bytecode and debuginfo *)
+      seek_in library_ichannel comp_unit.Cmo_format.cu_pos;
+      comp_unit.Cmo_format.cu_pos <- pos_out unit_ochannel;
+      copy_file_chunk library_ichannel unit_ochannel comp_unit.Cmo_format.cu_codesize;
+
+      if comp_unit.Cmo_format.cu_debug > 0 then begin
+        seek_in library_ichannel comp_unit.Cmo_format.cu_debug;
+        comp_unit.Cmo_format.cu_debug <- pos_out unit_ochannel;
+        copy_file_chunk library_ichannel unit_ochannel comp_unit.Cmo_format.cu_debugsize
+      end;
+
+      let pos_info = pos_out unit_ochannel in
+      (* writing compilation unit *)
+      output_value unit_ochannel comp_unit;
+      (* getting back to right after magic number and writing compilation_unit position *)
+      seek_out unit_ochannel ofs_pos_info;
+      output_binary_int unit_ochannel pos_info;
+    | CMXUnit (unit_info, crc) ->
+      output_string unit_ochannel Config.cmx_magic_number;
+      output_value unit_ochannel unit_info;
+      Digest.output unit_ochannel crc
+  );
+  close_out unit_ochannel
+;;
+
+(* Extract contents of the library. Units may have .cmo or .cmx extension,
+ * or have no extension at all. *)
+let extract_library_contents ichannel library_info units_to_extract =
+  let library_units = List.sort_uniq compare (unit_names_of_library library_info) in
+  let units_to_extract = List.sort_uniq compare units_to_extract in
+  let missing_in_library = find_missing_units library_units units_to_extract in
+  match missing_in_library with
+  | [] -> 
+    List.iter (extract_unit (ichannel, library_info)) units_to_extract
+  | _ ->
+    Printf.eprintf "Modules ";
+    List.iter (Printf.eprintf "%s ") missing_in_library;
+    Printf.eprintf "are not in the library.\n";
+    exit (-1)
```

Comment:
```
$ ./cmar -extract ../stdlib/stdlib.cma StdLabels Stack SSSSSS
Problem opening library file ../stdlib/stdlib.cma.
```

I think the above running code is supposed to reach this case, and prints `Modules SSSSSS are not in the library`, but it doesn't. The error is caught in `list_extract_library_units` instead.